### PR TITLE
Add vertex snippets

### PR DESCRIPTION
**Pull request by morganchen12**

### CODE REVIEW

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    let harassmentSafety = SafetySetting(harmCategory: .harassment, threshold: .blockOnlyHigh)
+    let hateSpeechSafety = SafetySetting(harmCategory: .hateSpeech, threshold: .blockMediumAndAbove)
+
+    let model = VertexAI.vertexAI().generativeModel(
+      modelName: "MODEL_NAME",
+      safetySettings: [harassmentSafety, hateSpeechSafety]
+    )
+    // [END set_safety_settings]
+  }
+
```
[the remark this answers is not in the document]
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:297]: I'll do function calling snippets in a follow-up.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+      print(text)
+    }
+    // [END multimodal_prompt]
+  }
+
+  func multiImagePromptStreaming() async throws {
+    // [START two_image_prompt_streaming]
+    #if canImport(UIKit)
+      guard let image1 = UIImage(named: "image1") else { fatalError() }
+      guard let image2 = UIImage(named: "image2") else { fatalError() }
```
[the remark this answers is not in the document]
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:143]: These snippets are just compiling, not running. So I'm ok with leaving them as-is despite the fact that when run they'll immediately encounter the `fatalError`.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    // To generate text output, call generateContent and pass in the prompt
+    let response = try await model.generateContent(image1, image2, prompt)
+    if let text = response.text {
+      print(text)
+    }
+    // [END two_image_prompt]
+  }
+
+  func textAndVideoPrompt() async throws {
+    // AVFoundation support coming soon™
```
[the remark this answers is not in the document]
Comment (andrewheard) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:184]: These LGTM too.

**Review by andrewheard — APPROVED** [review]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+    // Initialize the generative model with a model that supports your use case
+    // Gemini 1.5 Pro is versatile and can accept both text-only or multimodal prompt inputs
+    let model = vertex.generativeModel(modelName: "gemini-1.5-pro-preview-0409")
+    // [END initialize_model]
+
+    self.model = model
+  }
+
+  func callGemini() async throws {
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:42]: I think this is just a dup of the text-gen-from-text snippet, right?  If so, I don't think we need this.  On the getting started page, we'll just use the text-gen-from-text snippet.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+    // To generate text output, call generateContent with the text input
+    let response = try await model.generateContent(prompt)
+    if let text = response.text {
+      print(text)
+    }
+    // [END call_gemini]
+  }
+
+  func callGeminiStreaming() async throws {
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:55]: I think this is just a dup of the text-gen-from-text-streaming snippet, right? If so, I don't think we need this. On the getting started page, we'll just use the text-gen-from-text-streaming snippet.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    for try await chunk in contentStream {
+      if let text = chunk.text {
+        print(text)
+      }
+    }
+    // [END call_gemini_streaming]
+  }
+
+  func sendTextOnlyPromptStreaming() async throws {
+    // [START text_only_prompt_streaming]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:71]: [START text_only_prompt_streaming] => [START text_gen_text_only_prompt_streaming]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+    let contentStream = model.generateContentStream(prompt)
+    for try await chunk in contentStream {
+      if let text = chunk.text {
+        print(text)
+      }
+    }
+    // [END text_only_prompt_streaming]
+  }
+
+  // Note: This is the same as the call gemini prompt, but may change in the future.
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:85]: I think this is the one that we want to keep and remove the "call gemini" one above.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+  func sendTextOnlyPromt() async throws {
+    // [START text_only_prompt]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:87]: [START text_only_prompt] => [START text_gen_text_only_prompt]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+    // To generate text output, call generateContent with the text input
+    let response = try await model.generateContent(prompt)
+    if let text = response.text {
+      print(text)
+    }
+    // [END text_only_prompt]
+  }
+
+  func sendMultimodalPromptStreaming() async throws {
+    // [START multimodal_prompt_streaming]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:100]: [START multimodal_prompt_streaming] => [START text_gen_multimodal_one_image_prompt_streaming]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+    for try await chunk in contentStream {
+      if let text = chunk.text {
+        print(text)
+      }
+    }
+    // [END multimodal_prompt_streaming]
+  }
+
+  func sendMultimodalPrompt() async throws {
+    // [START multimodal_prompt]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:121]: [START multimodal_prompt] => [START text_gen_multimodal_one_image_prompt]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+    // To generate text output, call generateContent and pass in the prompt
+    let response = try await model.generateContent(image, prompt)
+    if let text = response.text {
+      print(text)
+    }
+    // [END multimodal_prompt]
+  }
+
+  func multiImagePromptStreaming() async throws {
+    // [START two_image_prompt_streaming]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:140]: [START two_image_prompt_streaming] => [START text_gen_multimodal_multi_image_prompt_streaming]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+    for try await chunk in contentStream {
+      if let text = chunk.text {
+        print(text)
+      }
+    }
+    // [END two_image_prompt_streaming]
+  }
+
+  func multiImagePrompt() async throws {
+    // [START two_image_prompt]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:163]: [START two_image_prompt] => [START text_gen_multimodal_multi_image_prompt]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+    let response = try await model.generateContent(image1, image2, prompt)
+    if let text = response.text {
+      print(text)
+    }
+    // [END two_image_prompt]
+  }
+
+  func textAndVideoPrompt() async throws {
+    // AVFoundation support coming soon™
+    // [START text_video_prompt]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:185]: [START text_video_prompt] => [START text_gen_multimodal_video_prompt]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+    // To generate text output, call generateContent and pass in the prompt
+    let response = try await model.generateContent(videoContent, prompt)
+    if let text = response.text {
+      print(text)
+    }
+    // [END text_video_prompt]
+  }
+
+  func textAndVideoPromptStreaming() async throws {
+    // [START text_video_prompt_streaming]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:201]: [START text_video_prompt_streaming] => [START text_gen_multimodal_video_prompt_streaming]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+
+    // To generate text output, call sendMessage and pass in the message
+    let response = try await chat.sendMessage("How many paws are in my house?")
+    if let text = response.text {
+      print(text)
+    }
+    // [END chat]
+  }
+
+  func countTokensText() async throws {
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:258]: question on these count tokens and billable chars snippets... can they be embedded inside each of the main snippets above?  With region tags added?

On the [count tokens docs page](https://firebase.google.com/docs/vertex-ai/count-tokens?platform=ios), I'd like to be able to show count tokens in the context of the other code.  And using special includecode syntax, there's a way to hide specific parts of a code snippet.  So, I can hide the count tokens bit in the main pages of the guides and only expose it in the count tokens page.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+    let message = try ModelContent(role: "user", "Why is the sky blue?")
+    let contents = history + [message]
+    let response = try await model.countTokens(contents)
+    print("Total Tokens: \(response.totalTokens)")
+    print("Total Billable Characters: \(response.totalBillableCharacters)")
+    // [END count_tokens_chat]
+  }
+
+  func setSafetySetting() {
+    // [START set_safety_setting]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:307]: [START set_safety_setting] => [START set_one_safety_setting]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+      modelName: "MODEL_NAME",
+      safetySettings: [
+        SafetySetting(harmCategory: .harassment, threshold: .blockOnlyHigh)
+      ]
+    )
+    // [END set_safety_setting]
+  }
+
+  func setMultipleSafetySettings() {
+    // [START set_safety_settings]
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:318]: [START set_safety_settings] => [START set_multi_safety_settings]

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+
+  var model: GenerativeModel!
+
+  func initializeModel() {
+    // [START initialize_model]
+    // Initialize the Vertex AI service
+    let vertex = VertexAI.vertexAI()
+
+    // Initialize the generative model with a model that supports your use case
+    // Gemini 1.5 Pro is versatile and can accept both text-only or multimodal prompt inputs
```
Comment (rachelsaunders) [inline VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:35]: // Gemini 1.5 models are versatile and can be used with all API capabilities

**Review by rachelsaunders — COMMENTED** [review]
thank you SO much for getting these into the snippet registry so quickly :-D

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 18 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:35]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 4 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:42]: Removed.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 5 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:55]: Removed.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 6 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:71]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 7 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:85]: Ack

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 8 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:87]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 9 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:100]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 10 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:121]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 11 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:140]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 12 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:163]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 13 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:185]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 14 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:201]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 15 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:258]: I don't think there's a way to nest snippets inside the exclude syntax, but you could certainly put them side by side.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 16 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:307]: Done.

File: VertexAISnippets/VertexAISnippets/VertexAISnippets.swift (hunk 17 above)
Comment (morganchen12) [reply VertexAISnippets/VertexAISnippets/VertexAISnippets.swift:318]: Done.

**Pull request merged**